packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaEditorUtils.ts

```diff
@@ -245,78 +258,95 @@ export function canHaveRestrictedValues(
+      if (columnModel.jsonSubColumns) {
+        columnModel.jsonSubColumns = (
+          columnModel.jsonSubColumns as JsonSubColumnModelFormData[]
+        ).map(
+          (
+            jsonSubColumnFormData: JsonSubColumnModelFormData,
+          ): JsonSubColumnModel => {
+            // isSelected field from the subcolumn for the same reason
+            const { isSelected, ...rest } = jsonSubColumnFormData
+            return {
+              ...rest,
+            }
+          },
+        )
       }
-    }
-    newSchema.push(copy)
-  }
-  const createdColumnModels = (await createColumnModels(accessToken, newSchema))
-    .list
+      return columnModel
+    },
+  )
+}
 
-  // now that all columns have been created, figure out the column changes (create, update, and no-op)
-  // keep track of column ids to figure out what columns were deleted.
-  const changes: ColumnChange[] = []
-  const columnIds: Set<string | null> = new Set()
-  for (let i = 0; i < proposedSchema.length; i++) {
-    const oldColumnId: string | null = proposedSchema[i].id ?? null
-    const newColumnId: string | null = createdColumnModels[i].id ?? null
-    columnIds.add(oldColumnId)
-    columnIds.add(newColumnId)
-    if (oldColumnId !== newColumnId) {
-      changes.push({ oldColumnId, newColumnId })
+/**
+ * Transform ColumnModels returned by Synapse into the form data for the TableColumnSchemaForm.
+ * Default column models, if available, are used to determine which columns should not be editable in the form.
+ * @param columnModels
+ * @param defaultColumns
+ */
+export function transformColumnModelsToFormData(
+  columnModels: ColumnModel[],
+  defaultColumns: ColumnModel[] = [],
+): ColumnModelFormData[] {
+  return columnModels.map((cm): ColumnModelFormData => {
+    return {
+      ...cm,
+      isSelected: false,
+      // If the name matches a known default column model, then we consider it to be a default column model itself
+      isOriginallyDefaultColumn: defaultColumns.some(
+        dcm => dcm.name === cm.name,
+      ),
+      jsonSubColumns: cm.jsonSubColumns
+        ? cm.jsonSubColumns.map(jsc => ({
+            ...jsc,
+            isSelected: false,
+          }))
+        : undefined,
     }
-  }
+  })
+}
 
-  // delete columns that are not represented in the changes already (create or update)
-  for (const oldColumnModel of oldSchema) {
-    const oldColumnId: string | null = oldColumnModel.id
-    if (!columnIds.has(oldColumnId)) {
-      changes.push({ oldColumnId, newColumnId: null })
+export function getViewScopeForEntity(entity: Entity): ViewScope | undefined {
+  if (isEntityView(entity)) {
+    return {
+      scope: entity.scopeIds,
+      viewTypeMask: entity.viewTypeMask,
+      viewEntityType: convertToEntityType(
+        entity.concreteType,
+      ) as ViewEntityType,
+    }
+  } else if (isDataset(entity) || isDatasetCollection(entity)) {
+    const mask = isDataset(entity)
```

Comment:
Is this logic reversed?  If isDataset(entity), shouldn't it use ENTITY_VIEW_TYPE_MASK_DATASET?

Reply:
No, the mask represents what the table contains. Datasets contain Files, and DatasetCollections contain Datasets.

Reply:
Oh, right!